crates/biome_js_analyze/src/lint/nursery/no_label_without_control.rs

```diff
@@ -0,0 +1,290 @@
+                markup! { "Consider adding a `for` or `htmlFor` attribute to the label element or moving the control element to inside the label element." },
+            );
+        }
+
+        Some(diagnostic)
+    }
+}
+
+/// Returns an array option merged with the default value
+fn get_option(value: &[String], default_value: &[&str]) -> Vec<String> {
+    value
+        .iter()
+        .cloned()
+        .chain(default_value.iter().map(|value| (*value).to_string()))
+        .collect()
+}
+
+/// Returns the `JsxAttributeList` of the passed `AnyJsxTag`
+fn get_element_attributes(jsx_tag: &AnyJsxTag) -> Option<JsxAttributeList> {
+    match jsx_tag {
+        AnyJsxTag::JsxElement(element) => Some(element.opening_element().ok()?.attributes()),
+        AnyJsxTag::JsxSelfClosingElement(element) => Some(element.attributes()),
+        _ => None,
+    }
+}
+
+/// Returns the element name of a `AnyJsxTag`
+fn get_element_name(jsx_tag: &AnyJsxTag) -> Option<String> {
+    match jsx_tag {
+        AnyJsxTag::JsxElement(element) => Some(element.opening_element().ok()?.name().ok()?.text()),
+        AnyJsxTag::JsxSelfClosingElement(element) => Some(element.name().ok()?.text()),
+        _ => None,
+    }
+}
+
+/// Returns whether the passed `AnyJsxTag` have a `for` or `htmlFor` attribute
+fn has_for_attribute(jsx_tag: &AnyJsxTag) -> Option<bool> {
+    let for_attributes = &["for", "htmlFor"];
+    let attributes = get_element_attributes(jsx_tag)?;
+
+    Some(attributes.into_iter().any(|attribute| {
+        match attribute {
+            AnyJsxAttribute::JsxAttribute(jsx_attribute) => jsx_attribute
+                .name()
+                .is_ok_and(|jsx_name| for_attributes.contains(&jsx_name.text().as_str())),
+            _ => false,
+        }
+    }))
+}
+
+/// Returns whether the passed `AnyJsxTag` have a child that is considered a control component
+/// according to the passed `control_components` parameter
+fn has_nested_control(jsx_tag: &AnyJsxTag, control_components: &[String]) -> bool {
+    jsx_tag.syntax().descendants().any(|descendant| {
+        match (
+            JsxName::cast(descendant.clone()),
+            JsxReferenceIdentifier::cast(descendant.clone()),
+        ) {
+            (Some(jsx_name), _) => control_components.contains(&jsx_name.text()),
+            (_, Some(jsx_reference_identifier)) => {
+                control_components.contains(&jsx_reference_identifier.text())
+            }
+            _ => false,
+        }
+    })
+}
+
+/// Returns whether the passed `AnyJsxTag` meets one of the following conditions:
+/// - Has a label attribute that corresponds to the `label_attributes` parameter
+/// - Has an `aria-labelledby` attribute
+/// - Has a child `JsxText` node
+fn has_accessible_label(jsx_tag: &AnyJsxTag, label_attributes: &[String]) -> bool {
+    let jsx_attributes = jsx_tag
+        .syntax()
+        .descendants()
+        .filter_map(JsxAttribute::cast)
+        .collect::<Vec<JsxAttribute>>();
```

Comment:
I don't think you need to collect here. `filter_map` still returns an iterator, so you can use `any` without collecting and allocating a vector.

Reply:
done!